chore: create basic layout components and associated utils

prsidhu opened the pull request:
## Description

Objective - Create scaffolding for layout components (LC).

- Created ``LayoutComponentHOC`` that will wrap the ``FlexLayout`` component and provide all the necessary functionality that is expected of an LC.
- Created ``LayoutFactory`` to initial available LCs.
- Created base LCs that will be used to construct presets - ``AlignedColumn, AlignedRow, Column and Row``.
- Created various utils to manage functionality of LCs.
- Added unit tests for utils.

#### PR fixes following issue(s)
Fixes #27000
Fixes #27001 
Fixes #27002 


#### Type of change
> Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)

## Testing

#### How Has This Been Tested?
> Please describe the tests that you ran to verify your changes. Also list any relevant details for your test configuration.
> Delete anything that is not relevant
- [x] Manual
- [ ] JUnit
- [x] Jest
- [ ] Cypress

## Checklist:
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] PR is being merged under a feature flag


#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

prsidhu commented:
/ok-to-test

marks0351 commented:
Some points of discussion wrt to code design:
To enable layout components inside the existing canvas flow i.e, the renderAppsmithCanvas we would need a Parent Component per each renderMode that would kick start rendering of layout components with the `layout` prop. 

`AnvilLayoutEditorCanvas` and `AnvilLayoutViewerCanvas`.
in both components an entry point is needed.

I see that renderLayouts is the kick stater which iterates through `layout` data of a canvas and gets the layout component based on type and renders it.
each layout component is wrapped with a hoc that adds a DnD layer and takes the decision whether the component should render layouts or render its own implementation of renderChildWidgets

Registration of layout components like widgets has hurdles coz we wont have access to app state during the registration cycle so doing it on-demand would not be possible with todays arch. I would recommend we skip the same widget registration flow and instead just have a constant mapping layout type and the component.

for the entry point mentioned above for AnvilLayoutEditorCanvas, Would it make sense to use a parent Component `LayoutComponetsMaker`, to pick the layout component from a constant which would be a map of all layout components which is mentioned in the `layout` property and render it instead of picking it from a layout registry that adds a HOC for every widget.

regarding the static functions per layout component, if they are to DRY the components, 
I would suggest we use an Abstract base class for Layout components(`BaseLayoutComponent`) just like BaseWidget and extend them per component with their own definitions.

The `LayoutComponentHOC` currently adds a DnD layer and the decision to renderLayouts or widgets, we can move this as a common logic inside the abstract class(`BaseLayoutComponent`) for layout components.
A bit skeptical about the effect of Object.assign in a HOC.

one more concern was that I feel a `LayoutComponent` is trying to do a lot by being a renderer of layout skeleton as well as being a renderer of children in a particular layout manner. there by the prop `layout` has multiple type definitions, I think we should isolate them, like have layout skeleton renderers and layout with widget renderers. and have appropriate type definitions and components for each. and whenever we see an opportunity to dry them we have tools like hooks, hocs and utils to do that.

@prsidhu @riodeuno let me know ur thoughts.

riodeuno commented on app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:
```diff
@@ -0,0 +1,164 @@
+export function getChildrenMap(
+  layoutProps: LayoutComponentProps,
+  map: LayoutComponentProps["childrenMap"],
+  res: LayoutComponentProps["childrenMap"] = {},
```
It seems like this third argument is the return value. Is there a reason why this exists as an argument?

riodeuno replied:
Please ignore this. Just saw the recursion in the function

riodeuno commented on app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:
```diff
@@ -0,0 +1,164 @@
+      // add widget to the resultant map.
+      res[each] = map[each];
+    } else if (Array.isArray(each)) {
+      // if type === array => each could be a list of widgetIDs.
```
Is this assumption valid? What if we have deeply nested layouts?

riodeuno commented on app/client/src/layoutSystems/anvil/utils/layouts/renderUtils.tsx:
```diff
@@ -0,0 +1,164 @@
+  const hasFillWidget: boolean = isFillWidgetPresentInList(
+    Object.values(childrenMap || {}),
+  );
+  const mode: RenderMode = renderMode || RenderModes.CANVAS;
```
Once again, I believe `renderMode` shouldn't be a part of props here in the system. Ideally, this should be a modal property of a higher context of the application.

riodeuno reviewed (commented):
I have reviewed much of the code, but not all of it. Will complete it by tomorrow.

prsidhu commented on app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:
```diff
@@ -1,12 +1,56 @@
 export interface LayoutComponentProps {
   layout: LayoutComponentProps[] | string[] | string[][]; // Array of layout components or widgets to render.
   layoutId: string; // Identifier of layout
   layoutStyle?: { [key: string]: any }; // React.CSSProperties for overriding default layout style.
```
@riodeuno yes, CSSProperties doesn't cover all scenarios. We also have to account for responsive values that are entertained by the container queries implementation on WDS.

prsidhu commented on app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:
```diff
@@ -1,12 +1,56 @@
-  layoutType: string; // Used to identify the correct layout component to render.
+  layoutType: LayoutComponentType; // Used to identify the correct layout component to render.
 
   allowedWidgetTypes?: string[]; // Array of widget types that can be dropped on the layout component.
```
empty array also signifies absence of a rule, which does the same job.

prsidhu commented on app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:
```diff
@@ -1,12 +1,56 @@
+    children: string[] | LayoutComponentProps[],
+    highlight: HighlightInfo,
+  ) => LayoutComponentProps;
+  getWidth: (arg0: any) => number;
```
@riodeuno @marks0351 Final definition and requirement of this function will become clear after the highlights task is done. I'll remove it if it doesn't serve a purpose anymore.

prsidhu commented on app/client/src/layoutSystems/anvil/utils/anvilTypes.ts:
```diff
@@ -1,12 +1,56 @@
+  // add all other static props here
+  type: LayoutComponentType;
+  // Add a child widget / layout to the parent layout component.
+  addChild: (
```
@riodeuno child can be either widget or layout. ``HighlightInfo`` provides the drop information that is used by the layout to place the child.

prsidhu commented on app/client/src/layoutSystems/anvil/utils/layouts/layoutUtils.ts:
```diff
@@ -0,0 +1,113 @@
+  const { rowIndex: index } = highlight;
+  return {
+    ...props,
+    layout: [...layout.slice(0, index), ...children, ...layout.slice(index)],
```
@riodeuno additionally, the layout will have to sort the children based on the said ``index`` for every render, as well as highlights calculation.

prsidhu commented:
@riodeuno @marks0351 I have added the canvas implementation to this PR, along with minor review suggestions. 

I will include the larger suggestions in the anvil sagas PR as it will allow me to make the desired changes across the board in a single attempt.

Additionally, I've added three tasks to reconsider epic based on the review suggestions.

prsidhu commented:
/ok-to-test

riodeuno commented:
/ok-to-test

prsidhu commented:
Closing this. The changes have been merged to release via #27980
